### PR TITLE
[TTS] Create skeleton for TTS inference API

### PR DESCRIPTION
**Pull request by rlangman**
# What does this PR do ?

Add prototype implementation of TTS pipeline that is intended to be used for inference before and after model training.

These are far from finalized. The intent is for this to be experimental and to slowly incorporate them into our inference workflows (eg. for preprocessing, evaluation, analysis, deployment) and expand and iterate on the interfaces and design as we figure out what needs to be changed.

**Collection**: [TTS]

# Changelog
- Add TTS pipeline and module definitions and documentation.
- Add module implementations using existing backends.
- Add unit and functional tests for all modules. 

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: nemo/collections/tts/inference/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        Args:
+            text: post-processed text string.
+
+        Returns:
+            phoneme string
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def convert_graphemes_to_phonemes_mixed(self, text: str, phone_prob: float) -> str:
```
Comment (redoctopus) [inline nemo/collections/tts/inference/pipeline.py:81]: Why have both this and the previous `convert_graphemes_to_phonemes()` if setting `phone_prob` will cover both cases?

**Review by redoctopus — COMMENTED** [review]
Thanks for starting this API!

File: nemo/collections/tts/inference/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        Args:
+            text: input text string.
+
+        Returns:
+            list of integer indices.
+        """
+        raise NotImplementedError()
+
+
+class AudioProcessor(ABC):
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/inference/pipeline.py:110]: Right now all of the module interfaces (abstract classes) are defined in pipeline.py with possible implementations in other files.

Alternatively, I could put the interfaces in the same python file as their implementations. This would make it a little more difficult to see how all the modules fit together, but would make it easier to organize as the number of interfaces grows.

File: nemo/collections/tts/inference/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+
+class SpectrogramSynthesizer(ABC):
+    @property
+    @abstractmethod
+    def device(self):
+        pass
+
+    @abstractmethod
+    def synthesize_spectrogram(self, tokens: torch.tensor, speaker: torch.tensor, pitch: torch.tensor) -> torch.tensor:
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/inference/pipeline.py:134]: That would make sense as it is in our current fastpitch export. Should the pace be a scalar that speeds up or slows down the entire utterance, or the same as pitch with 1 pace value per token? The fastpitch code only supports the former but the latter would make more sense to me.

Also, it would not necessarily be an "option". The functions from tensor to tensor should not have optional arguments so they can be exported. The pipeline class could however handle the logic of creating default inputs for the user (eg. set pitch to duplicated 0.0 and pace to 1.0)

File: nemo/collections/tts/inference/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+            text: text string to synthesize.
+            speaker: optional integer speaker ID of voice to synthesize, default 0.
+            pitch: optional float pitch adjustment for synthesized utterance, default 0.
+
+        Returns:
+            [T_audio] numpy array containing synthesized audio samples.
+        """
+
+        assert self.spectrogram_synthesizer is not None
+        assert self.vocoder is not None
```
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/inference/pipeline.py:247]: I am not sure at this point. Probably E2E will require a separate pipeline definition. Some methods like get_spectrogram() will not be relevant, either.

File: nemo/collections/tts/inference/pipeline.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/inference/pipeline.py:110]: Do you think it would make sense to tack "`_interfaces`" on as a suffix for the relevant files? I like the separation between interfaces/implementations but just want to make sure the files are easy to navigate at a glance.

File: nemo/collections/tts/inference/pipeline.py (hunk 1 above)
Comment (redoctopus) [reply nemo/collections/tts/inference/pipeline.py:81]: Gotcha, sounds good.

File: nemo/collections/tts/inference/pipeline.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rlangman) [reply nemo/collections/tts/inference/pipeline.py:110]: Like, have both a pipeline.py and pipeline_interfaces.py? Or split the interfaces into multiple files?

[the remark this answers is not in the document]
Comment (redoctopus) [reply nemo/collections/tts/inference/pipeline.py:110]: Maybe something like `pipeline_interfaces.py` with the other files where they are now?

**Review by redoctopus — DISMISSED** [review]
LGTM, @XuesongYang any thoughts?

File: nemo/collections/tts/inference/audio_processors.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/audio_processors.py:1]: `NVIDIA CORPORATION & AFFILIATES`

File: nemo/collections/tts/inference/inference_modules.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/inference_modules.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: tests/collections/tts/inference/test_audio_processors.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline tests/collections/tts/inference/test_audio_processors.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: nemo/collections/tts/inference/vocoders.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/vocoders.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: nemo/collections/tts/inference/text_processors.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/text_processors.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: nemo/collections/tts/inference/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/pipeline.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: nemo/collections/tts/inference/spectrogram_synthesizers.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/spectrogram_synthesizers.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: tests/collections/tts/inference/test_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline tests/collections/tts/inference/test_pipeline.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: tests/collections/tts/inference/test_spectrogram_synthesizers.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline tests/collections/tts/inference/test_spectrogram_synthesizers.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: tests/collections/tts/inference/test_text_processors.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline tests/collections/tts/inference/test_text_processors.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: tests/collections/tts/inference/test_vocoders.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (XuesongYang) [inline tests/collections/tts/inference/test_vocoders.py:1]: 2023, NVIDIA CORPORATION & AFFILIATES.

File: nemo/collections/tts/inference/inference_modules.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        Args:
+            text: post-processed text string.
+
+        Returns:
+            phoneme string
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def convert_graphemes_to_phonemes_mixed(self, text: str, phone_prob: float) -> str:
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/inference_modules.py:78]: it seems we can make a default value for `phone_prob` in `convert_graphemes_to_phonemes` such that `convert_graphemes_to_phonemes_mixed` can be removed.

File: nemo/collections/tts/inference/pipeline.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+        vocoder: Vocoder = None,
+    ):
+        self.text_processor = text_processor
+        self.g2p = g2p
+        self.text_tokenizer = text_tokenizer
+        self.audio_processor = audio_processor
+        self.spectrogram_synthesizer = spectrogram_synthesizer
+        self.vocoder = vocoder
+
+    def get_spectrogram(self, audio: np.ndarray) -> np.ndarray:
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/pipeline.py:47]: any reason why we use `np.ndarray` instead of `torch.tensor`?

File: nemo/collections/tts/inference/pipeline.py [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+    SpectrogramSynthesizer,
+    TextProcessor,
+    TextTokenizer,
+    Vocoder,
+)
+from nemo.utils.decorators import experimental
+
+
+@experimental
+class TTSPipeline:
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/pipeline.py:30]: Maybe we should make a clear name for this class and distinguish e2e from cascaded ones. If you remember, we now have both two-stage like TTS pipeline (text -> mel-gen -> audio) and an end-to-end pipeline (text -> audio).

File: tests/collections/tts/inference/test_spectrogram_synthesizers.py [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+from nemo.collections.tts.inference.spectrogram_synthesizers import FastPitchSpectrogramSynthesizer
+from nemo.collections.tts.models import FastPitchModel
+
+
+class TestSpectrogramSynthesizers(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super(TestSpectrogramSynthesizers, cls).setUpClass()
+
+        fastpitch_model = FastPitchModel.from_pretrained("tts_en_fastpitch_multispeaker").eval().to("cpu")
```
Comment (XuesongYang) [inline tests/collections/tts/inference/test_spectrogram_synthesizers.py:29]: it is fine to only download one ngc model and run evaluation here as a unit test, but I would suggest decorating it with a marker `nightly` and letting QA run the test. Our CI tried not to download NGC models for inference.

Comment (XuesongYang) [reply tests/collections/tts/inference/test_spectrogram_synthesizers.py:29]: see examples: https://github.com/NVIDIA/NeMo/tree/main/tests/collections/tts/models

Comment (XuesongYang) [reply tests/collections/tts/inference/test_spectrogram_synthesizers.py:29]: besides, be aware of the instance's life cycle for downloading the NGC model. Try not to duplicate downloading the NGC. `@fixture` decor could help.

File: nemo/collections/tts/inference/text_processors.py [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+class BaseTextProcessor(TextProcessor):
+    def __init__(self, normalizer: Normalizer):
+        self.normalizer = normalizer
+
+    def preprocess_text(self, text: str) -> str:
+        return text
+
+    def normalize_text(self, text: str) -> str:
+        normalized_text = self.normalizer.normalize(text, punct_pre_process=True, punct_post_process=True)
```
Comment (XuesongYang) [inline nemo/collections/tts/inference/text_processors.py:31]: IIRC, our normalization function call doesn't support multi-processing. So we may need to enforce a multi-processing in this function for better computing efficiency.

Comment (XuesongYang) [reply nemo/collections/tts/inference/text_processors.py:31]: FYI, i used this before: https://github.com/NVIDIA/NeMo/blob/main/scripts/dataset_processing/tts/hui_acg/get_data.py#L201-L202

Comment (XuesongYang) [reply nemo/collections/tts/inference/text_processors.py:31]: a second look, it seems your design only takes as input a single `str` while not a batch. Maybe the multiprocessing should be moved to `pipeline.py`.

File: tests/collections/tts/inference/test_spectrogram_synthesizers.py (hunk 19 above)
Comment (rlangman) [reply tests/collections/tts/inference/test_spectrogram_synthesizers.py:26]: Added `@pytest.mark.nightly` and `@fixture` for all tests that download pretrained models.
